kyo-core/shared/src/main/scala/kyo/Async.scala

```diff
@@ -0,0 +1,368 @@
+import scala.annotation.targetName
+import scala.collection.immutable.ArraySeq
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+import scala.util.NotGiven
+import scala.util.control.NonFatal
+import scala.util.control.NoStackTrace
+
+opaque type Async <: (IO & Async.Join) = Async.Join & IO
+
+object Async:
+
+    sealed trait Join extends Effect[IOPromise[?, *], Result[Nothing, *]]
+
+    inline def run[E, A: Flat, Ctx](inline v: => A < (Abort[E] & Async & Ctx))(
+        using
+        boundary: Boundary[Ctx, IO],
+        reduce: Reducible[Abort[E]],
+        frame: Frame
+    ): Fiber[E, A] < (IO & Ctx) =
+        boundary((trace, context) => IOTask(v, trace, context))
+
+    def runAndBlock[E, A: Flat, Ctx](timeout: Duration)(v: => A < (Abort[E] & Async & Ctx))(
+        using
+        boundary: Boundary[Ctx, IO & Abort[E]],
+        frame: Frame
+    ): A < (Abort[E | Timeout] & IO & Ctx) =
+        run(v).map { fiber =>
+            IO(Abort.get(fiber.block(deadline(timeout))))
+        }
+    end runAndBlock
+
+    opaque type Promise[E, A] <: Fiber[E, A] = IOPromise[E, A]
+
+    object Promise:
+        inline given [E, A]: Flat[Promise[E, A]] = Flat.unsafe.bypass
+
+        def init[E, A](using Frame): Promise[E, A] < IO = IO(IOPromise())
+
+        extension [E, A](self: Promise[E, A])
+            def complete[E2 <: E, A2 <: A](v: Result[E, A])(using Frame): Boolean < IO     = IO(self.complete(v))
+            def completeUnit[E2 <: E, A2 <: A](v: Result[E, A])(using Frame): Unit < IO    = IO(discard(self.complete(v)))
+            def become[E2 <: E, A2 <: A](other: Fiber[E2, A2])(using Frame): Boolean < IO  = IO(self.become(other))
+            def becomeUnit[E2 <: E, A2 <: A](other: Fiber[E2, A2])(using Frame): Unit < IO = IO(discard(self.become(other)))
+        end extension
+    end Promise
+
+    opaque type Fiber[E, A] = IOPromise[E, A]
+
+    object Fiber extends FiberPlatformSpecific:
+
+        inline given [E, A]: Flat[Fiber[E, A]] = Flat.unsafe.bypass
+
+        private val _unit = success(())
+
+        def unit[E]: Fiber[E, Unit] = _unit.asInstanceOf[Fiber[E, Unit]]
+
+        def never: Fiber[Nothing, Unit] = IOPromise[Nothing, Unit]()
+
+        def success[E, A](v: A): Fiber[E, A]        = result(Result.success(v))
+        def fail[E, A](ex: E): Fiber[E, A]          = result(Result.fail(ex))
+        def panic[E, A](ex: Throwable): Fiber[E, A] = result(Result.panic(ex))
+
+        def fromFuture[A: Flat](f: Future[A])(using Frame): Fiber[Nothing, A] < IO =
+            import scala.util.*
+            IO {
+                val p = new IOPromise[Nothing, A]()
+                f.onComplete {
+                    case Success(v) =>
+                        p.complete(Result.success(v))
+                    case Failure(ex) =>
+                        p.complete(Result.panic(ex))
+                }(ExecutionContext.parasitic)
+                p
```

Comment:
It does but then the result would be a `Fiber[Throwable, A]` instead of `Fiber[Nothing, A]`. I'm honestly not sure what error scenarios we should track but tracking `Throwable` doesn't seem very valuable since `Panic` is essentially that.

Reply:
Given that `Future` tracks Throwables, I think it makes sense that should be tracked in the Fiber as well.

Panics should usually be reserved for unrecoverable errors, for which I don't think we expect from `Future`.

Reply:
There are a few other places where I didn't track exceptions as well, for example when `Closed` can be thrown in `Channel`. Could we create a "review tracked errors" ticket and address this separately?

Reply:
Sure that's fine